File: app/src/main/java/pl/schibsted/mateuszderks/findme/RoomsAdapter.java
package pl.schibsted.mateuszderks.findme;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;


import java.util.List;
import java.util.ArrayList;

import pl.schibsted.mateuszderks.findme.resource.RoomApi;
import retrofit.Callback;
import retrofit.RestAdapter;
import retrofit.RetrofitError;
import retrofit.client.Response;

public class RoomsAdapter extends RecyclerView.Adapter<ViewHolder> {
    List<Room> items = new ArrayList<Room>();

    RoomsAdapter() {
        RestAdapter restAdapter = new RestAdapter.Builder()
                .setEndpoint("https://agile-brushlands-2260.herokuapp.com/api/v1")
                .build();

        RoomApi roomApi = restAdapter.create(RoomApi.class);

        Callback<List<Room>> callback = new Callback<List<Room>>() {
            @Override
            public void success(List<Room> o, Response response) {
                addAll(o);
            }

            @Override
            public void failure(RetrofitError retrofitError) {

            }
        };

        roomApi.list(callback);
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
        View itemView = LayoutInflater
                .from(viewGroup.getContext())
                .inflate(R.layout.item_room, viewGroup, false);

        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(ViewHolder viewHolder, int i) {
        Room room = items.get(i);

        viewHolder.name.setText(room.name + " - " + room.id);
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public void addAll(List<Room> list) {
        items.addAll(list);
        notifyDataSetChanged();
    }
}

File: app/src/main/kotlin/pl/schibsted/mateuszderks/findme/RoomsAdapter.kt
package pl.schibsted.mateuszderks.findme

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup

import pl.schibsted.mateuszderks.findme.resource.RoomApi
import retrofit.RestAdapter
import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import android.view.View
import butterknife.bindView
import android.widget.TextView
import pl.schibsted.mateuszderks.findme.RoomsAdapter.ViewHolder
import java.util.concurrent.TimeUnit
import rx.schedulers.Timestamped
import android.text.format.DateFormat

public class RoomsAdapter : RecyclerView.Adapter<ViewHolder>() {
    val items: MutableList<Room> = arrayListOf();

    {
        val restAdapter = RestAdapter
                .Builder()
                .setEndpoint("https://agile-brushlands-2260.herokuapp.com/api/v1")
                .build()

        val roomApi = restAdapter.create<RoomApi>(javaClass<RoomApi>())

        println("CONNECTION");

        roomApi
                .list()
                .flatMap { Observable.from(it) }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    items.add(0, it)
                    notifyItemInserted(0)
                }) {
                    println("Error ${it.getMessage()}")
                    it.printStackTrace()
                }
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): ViewHolder {
        val itemView = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.item_room, viewGroup, false)

        return ViewHolder(itemView);
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, i: Int) {
        val room = items.get(i);

        viewHolder.name.setText(room.name);
    }

    override fun getItemCount(): Int {
        return items.size();
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val name: TextView by bindView(R.id.room_name);
    }
}
